feat: add warning banner if bridge fees are over limit

cowdan opened the pull request:
# Summary

- Add a warning banner if bridge fees are over limit.
- Refactor the warning banners to be more aligned with slippage banner

Before:
<img width="497" height="645" alt="Screenshot 2025-07-22 at 09 41 30" src="https://github.com/user-attachments/assets/f377b1e5-5bd1-4e9c-a260-497f43994d2e" />

After:
<img width="627" height="1021" alt="Screenshot 2025-07-24 at 17 06 57" src="https://github.com/user-attachments/assets/0f9b6199-f485-40db-8274-8ad458ea66f4" />

<img width="628" height="1072" alt="Screenshot 2025-07-24 at 17 08 47" src="https://github.com/user-attachments/assets/f8fd36d1-0ccb-427e-984c-95dd3790d94d" />



# To Test

1. Prepare a swap and bridge with a minimum amount so that the fees are high. For example, sell 5 USDC on Mainnet  and buy 5 USDT on Polygon


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

* **New Features**
  * Unified high fee warning banner now covers both swap and bridge fees with dynamic messaging.
  * Added an embedded "Swap anyway" checkbox within the warning banner for logged-in users.
  * High fee warnings consistently displayed in the swap confirmation modal regardless of bridge details.

* **Improvements**
  * Enhanced banner layout and styling with improved tooltip and color coding based on fee severity.
  * Extended fee warning logic to include bridge fees and improved type safety in fee calculations.
  * InlineBanner component updated to support custom content layout for richer banner displays.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alfetopito commented on apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/HighFeeWarning/index.tsx:
```diff
@@ -35,27 +36,55 @@ export function HighFeeWarning({ readonlyMode }: HighFeeWarningProps) {
+          )
+        }
+      >
+        Costs exceed {level}% of the swap amount
```
This is what Anxo mentioned yesterday.
To be more transparent, we can display the actual percentage here.
The `feePercentage` is already returned by `useHighFeeWarning`, just need to use it here.

elena-zh commented:
Hey @cowdan , I have not done a comprehensive testing, however, found some issues: 
1. Not connected state, regular swap: there are 2 warnings instead of 1
<img width="663" height="788" alt="image" src="https://github.com/user-attachments/assets/403ecbdf-d276-4a63-8b8e-09ea9c5d43c3" />
2. Overall, I don't think we need to show 2 warnings. IMO, we should show only one: 
- 'combined' for the swap and bridge order. E.g. 'Costs exceed % of a swap and bridge amount'
- and a regular one for a swap in 1 chain
3. Same as case 1: 2nd warning should not be displayed for a regular trade
<img width="1715" height="710" alt="image" src="https://github.com/user-attachments/assets/5b8bc8db-a9dc-4eac-a591-7cf8cf6daa92" />


4. When I check the 1st box, the 2nd one is automatically checked as well 
<img width="553" height="235" alt="image" src="https://github.com/user-attachments/assets/5085cd10-7b12-4542-b831-34502492feb1" />

5. Banner should be smaller on the Confirm modal 
<img width="616" height="435" alt="image" src="https://github.com/user-attachments/assets/fc1acb2e-b3b3-4250-a4c2-2471d8c78eb7" />
See an example: 
<img width="697" height="233" alt="image" src="https://github.com/user-attachments/assets/32371074-1cab-482a-bd5f-9d02542d6c2c" />

6. Here a bridge fee > 20% of a sell amount, however, there is no banner 
<img width="657" height="812" alt="image" src="https://github.com/user-attachments/assets/e301c9a9-8cdf-44a4-a063-6c8caab19fa4" />

Here a fee is >50%, and still no banner 
<img width="565" height="773" alt="image" src="https://github.com/user-attachments/assets/c9b7e6f8-cc73-49b5-a9bf-3c5c963fd800" />

fairlighteth commented on apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/HighFeeWarning/index.tsx:
```diff
@@ -1,29 +1,30 @@
+}
+
 // TODO: Add proper return type annotation
 // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
might as well tackle this one as well here

cowdan commented on apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/HighFeeWarning/hooks/useHighFeeWarning.ts:
```diff
@@ -10,27 +11,36 @@ import { useSafeEffect, useSafeMemo } from 'common/hooks/useSafeMemo'
 
   // only considers inputAmount vs fee (fee is in input token)
-  const [isHighFee, feePercentage] = useMemo(() => {
+  const [isHighFee, feePercentage, isHighBridgeFee, bridgeFeePercentage] = useMemo(() => {
```
if there's a `bridgeFee`, the value of `isSell` is always `true` for business reasons

elena-zh commented:
@cowdan , could you please update the issue description? It does not match to the behavior that I see on the PR preview link

cowdan commented on apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/HighFeeWarning/index.tsx:
```diff
@@ -1,68 +1,97 @@
+    return bridgeLevel
+  }, [feePercentage, bridgeFeePercentage])
+
+  const textContent = useMemo(() => {
```
done

fairlighteth commented:
@elena-zh yes this is what I suggested in https://github.com/cowprotocol/cowswap/pull/6009#issuecomment-3103363998

cowdan commented:
> Hey @cowdan , thank you! Looks way better! Still somw notes and issues:
> 
> 1. Banner where both swap and bridge costs are high does not look good:

Should be ok now

> 2. In case swap costs are high, I would not add bridge fees info into the tooltip. E.g. it says 'consider waiting for lower network costs', however, it is not related to the bridge part
> 

Updated text content

"> 3. Same when bridge costs are high: I would not mention swap costs in the tooltip

Updated text content

> 4. Bridge costs percentage is calculated based on the swap amount, not a bridge amount--> it is wrong to say thar costs exceed a bridge amount

Updated text content

> @fairlighteth , I think this question is for you: can we say
> 
> * "Bridge costs are at least XX% of a swap amount' instead?

> * and 'Swap costs are at least XX% and bridge costs are at least YY% of the swap amount?

Updated this too

cowdan commented on apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/HighFeeWarning/highFeeWarningHelpers.ts:
```diff
@@ -0,0 +1,48 @@
+  }
+}
+
+const formatFeePercentage = (feePercentage: Fraction | undefined): string => {
+  return feePercentage?.toSignificant(2, undefined, Rounding.ROUND_DOWN) ?? '0'
+}
```
If the fee is 25.4%, saying `at least 26%` is incorrect

elena-zh commented:
Hey @cowdan , great, thank you!
Could you please return back this text into the bridge costs tooltip? 
<img width="459" height="221" alt="would not remove" src="https://github.com/user-attachments/assets/93ed2a2f-9bab-4c99-afdf-27821d97c795" />

<img width="392" height="98" alt="image" src="https://github.com/user-attachments/assets/2da43b01-a869-4d61-bdb0-a58baf8c1c8b" />

Thank you

cowdan commented:
> Hey @cowdan , great, thank you! Could you please return back this text into the bridge costs tooltip? 

This should be fixed now

elena-zh reviewed (approved):
Thank you so much!

fairlighteth commented on apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/HighFeeWarning/hooks/useHighFeeWarning.ts:
```diff
@@ -10,27 +11,36 @@ import { useSafeEffect, useSafeMemo } from 'common/hooks/useSafeMemo'
+export function useHighFeeWarning(): UseHighFeeWarningReturn {
   const receiveAmountInfo = useGetReceiveAmountInfo()
 
   const [feeWarningAccepted, setFeeWarningAccepted] = useAtom(feeWarningAcceptedAtom)
```
Are we OK with tracking a single value or do we need keyed entity identifiers for multi-entity data?

cowdan replied:
a single value is enough as we have only one banner that combines both swap and bridge warnings

fairlighteth commented on apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/HighFeeWarning/HighFeeWarningTooltipContent.tsx:
```diff
@@ -0,0 +1,78 @@
+  bridgeFeePercentage?: Fraction
+}
+
+const Wrapper = ({ children }: { children: React.ReactNode }): React.ReactNode => {
```
Should we convert all these render functions to proper React components

shoom3301 replied:
This is a valid React component function.
Anyway, code style should be fixed (`React.ReactNode` -> `ReactNode`)

cowdan replied:
right, done

cowdan commented on apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/HighFeeWarning/highFeeWarningHelpers.ts:
```diff
@@ -0,0 +1,48 @@
+import { HIGH_TIER_FEE, LOW_TIER_FEE, MEDIUM_TIER_FEE } from './consts'
+
+// checks fee as percentage (30% not a decimal)
+export function _getWarningInfo(feePercentage?: Fraction): 30 | 20 | 10 | undefined {
```
done
